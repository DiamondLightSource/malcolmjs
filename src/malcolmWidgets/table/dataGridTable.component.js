/* eslint react/no-array-index-key: 0 */
/* eslint no-underscore-dangle: 0 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize, fade } from '@material-ui/core/styles/colorManipulator';
import IconButton from '@material-ui/core/IconButton';
import Add from '@material-ui/icons/Add';
import { Typography } from '@material-ui/core';
import ReactDataGrid from 'react-data-grid';
import AttributeAlarm, {
  AlarmStates,
} from '../attributeDetails/attributeAlarm/attributeAlarm.component';
import TableWidgetSelector, { getTableWidgetTags } from './widgetSelector';

const styles = theme => ({
  headerLayout: {
    tableLayout: 'fixed',
    width: 'calc(100% - 15px - 32px)',
  },
  footerLayout: {
    width: '100%',
  },
  headerLayoutNoScroll: {
    tableLayout: 'fixed',
    width: 'calc(100% - 32px)',
  },
  tableLayout: {
    tableLayout: 'fixed',
  },
  tableBody: {
    overflowY: 'auto',
    height: 'calc(100% - 75px)',
    width: '100%',
    backgroundColor: theme.palette.background.paper,
  },
  rowFormat: {
    height: '36px',
  },
  incompleteRowFormat: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.5),
    textAlign: 'Center',
    padding: '2px',
  },
  textHeadings: {
    borderLeft: `2px solid ${theme.palette.divider}`,
    textAlign: 'Center',
  },
  selectedRow: {
    backgroundColor: fade(theme.palette.secondary.main, 0.25),
    textAlign: 'Center',
    padding: '2px',
  },
  textBody: {
    backgroundColor: theme.palette.background.default,
    textAlign: 'Center',
    padding: '2px',
  },
  button: {
    width: '24px',
    height: '24px',
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
});

const isSelectedRow = props =>
  props.selectedRow === props.row
    ? props.classes.selectedRow
    : props.classes.textBody;

const AlarmCell = props => (
  <IconButton
    className={props.classes.button}
    disableRipple
    onClick={() => {
      props.infoClickHandler(props.path, `row.${props.row}`);
    }}
  >
    <AttributeAlarm
      alarmSeverity={
        props.flags.rows[props.row] &&
        (props.flags.rows[props.row]._dirty ||
          props.flags.rows[props.row]._isChanged)
          ? AlarmStates.DIRTY
          : AlarmStates.NO_ALARM
      }
    />
  </IconButton>
);

class NewWidgetTable extends React.Component {
  componentDidMount() {
    const headerCells = [
      ...document.getElementsByClassName('react-grid-HeaderCell'),
    ];
    const header = [...document.getElementsByClassName('react-grid-Header')];
    const canvas = [...document.getElementsByClassName('react-grid-Canvas')];
    if (headerCells.length !== 0) {
      headerCells.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.paper
          }; font-family: roboto; border-bottom-width: 0; border-right-width: 2px; border-right-color: ${
            this.props.theme.palette.divider
          } height: 36px; padding: 0; vertical-align: middle; margin-right: 2px`
        );
      });
    }
    if (canvas.length !== 0) {
      canvas.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.default
          };`
        );
      });
    }
    if (header.length !== 0) {
      header.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.paper
          };`
        );
      });
    }
  }
  render() {
    const columnLabels =
      this.props.localState === undefined
        ? Object.keys(this.props.attribute.raw.meta.elements)
        : this.props.localState.labels;
    const values =
      this.props.localState === undefined
        ? this.props.attribute.raw.value[columnLabels[0]].map((val, row) => {
            const rowData = {};
            columnLabels.forEach(label => {
              rowData[label] = this.props.attribute.raw.value[label][row];
            });
            return rowData;
          })
        : this.props.localState.value;
    const flags =
      this.props.localState === undefined
        ? { rows: [], table: {} }
        : this.props.localState.flags;
    const meta =
      this.props.localState === undefined
        ? this.props.attribute.raw.meta
        : this.props.localState.meta;
    const rowChangeHandler = (rowPath, newValue) => {
      let rowValue;
      if (columnLabels === undefined) {
        rowValue = newValue;
      } else {
        rowValue = {};
        columnLabels.forEach(label => {
          rowValue[label] = values[rowPath.row][label];
          return 0;
        });
        rowValue[rowPath.label] = newValue;
      }
      this.props.eventHandler(
        this.props.attribute.calculated.path,
        rowValue,
        rowPath.row
      );
    };
    const rowFlagHandler = (rowPath, flagType, flagState) => {
      if (this.props.localState !== undefined) {
        const rowFlags =
          this.props.localState.flags.rows[rowPath.row] === undefined
            ? {}
            : this.props.localState.flags.rows[rowPath.row];
        if (rowFlags[flagType] === undefined) {
          rowFlags[flagType] = {};
        }
        rowFlags[flagType][rowPath.label] = flagState;
        rowFlags[`_${flagType}`] = Object.values(rowFlags[flagType]).some(
          val => val
        );
        this.props.setFlag(
          this.props.attribute.calculated.path,
          rowPath.row,
          flagType,
          rowFlags
        );
      }
    };
    const columnWidgetTags = getTableWidgetTags(meta);
    const columnHeadings =
      columnLabels === undefined
        ? [
            {
              key: 'value',
              name: meta.label,
              resizable: true,
            },
          ]
        : columnLabels.map(label => ({
            key: label,
            name: meta.elements[label].label
              ? meta.elements[label].label
              : label,
            resizable: true,
          }));
    if (!this.props.hideInfo) {
      columnHeadings.splice(0, 0, {
        key: 'alarmState',
        name: 'Alarm',
        resizable: true,
      });
    }
    const getRowData = (row, columnWidgetTags, props) => {
      let valueCells;
      if (columnLabels === undefined) {
        valueCells = {
          value: (
            <TableWidgetSelector
              columnWidgetTag={columnWidgetTags[0]}
              value={values[row]}
              rowPath={{ row, column: 0 }}
              rowChangeHandler={rowChangeHandler}
              columnMeta={meta}
              setFlag={rowFlagHandler}
            />
          ),
        };
      } else {
        valueCells = {};
        columnLabels.forEach((label, column) => {
          valueCells[label] = (
            <TableWidgetSelector
              columnWidgetTag={columnWidgetTags[column]}
              value={values[row][label]}
              rowPath={{ label, row, column }}
              rowChangeHandler={rowChangeHandler}
              columnMeta={meta.elements[label]}
              setFlag={rowFlagHandler}
            />
          );
        });
      }
      if (!props.hideInfo) {
        valueCells.alarmState = (
          <AlarmCell
            flags={flags}
            row={row}
            classes={props.classes}
            path={props.attribute.calculated.path}
            infoClickHandler={props.infoClickHandler}
            selectedRow={props.selectedRow}
          />
        );
      }
      return valueCells;
    };

    console.log('hello!');
    return (
      <div style={{ width: '100%', height: '100%' }}>
        <ReactDataGrid
          columns={columnHeadings}
          rowsCount={values.length}
          rowGetter={row => getRowData(row, columnWidgetTags, this.props)}
          minHeight={700}
        />
        {meta.writeable ? (
          <div>
            <IconButton
              onClick={() =>
                this.props.addRow(
                  this.props.attribute.calculated.path,
                  values.length
                )
              }
            >
              <Add style={{ width: '32px', height: '32px' }} />
            </IconButton>
          </div>
        ) : null}
        <div style={{ display: 'flex', position: 'bottom', padding: '4px' }}>
          {this.props.footerItems}
        </div>
      </div>
    );
  }
}
export default withStyles(styles, { withTheme: true })(NewWidgetTable);
