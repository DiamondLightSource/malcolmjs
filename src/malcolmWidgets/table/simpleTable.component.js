import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize } from '@material-ui/core/styles/colorManipulator';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

const styles = theme => ({
  rowFormat: {
    height: '30px',
  },
  textHeadings: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.3),
    textAlign: 'Center',
  },
  textBody: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.1),
    textAlign: 'Center',
  },
});

const SimpleTable = props => {
  const columnHeadings = props.labels.map(label => (
    <TableCell className={props.classes.textHeadings} padding="none">
      {label}
    </TableCell>
  ));
  const table = props.labels.map(label => props.values[label]);
  return (
    <Table>
      <TableHead>
        <TableRow className={props.classes.rowFormat}>
          {columnHeadings}
        </TableRow>
      </TableHead>
      <TableBody>
        {table[0].map((value, row) => (
          <TableRow className={props.classes.rowFormat}>
            {props.labels.map((label, column) => (
              <TableCell className={props.classes.textBody} padding="none">
                {table[column][row]}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
};

SimpleTable.propTypes = {
  labels: PropTypes.arrayOf(PropTypes.string).isRequired,
  values: PropTypes.shape({}).isRequired,
  classes: PropTypes.shape({
    textBody: PropTypes.string,
    textHeadings: PropTypes.string,
    rowFormat: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles, { withTheme: true })(SimpleTable);
