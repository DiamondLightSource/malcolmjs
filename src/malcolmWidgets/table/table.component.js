/* eslint react/no-array-index-key: 0 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize } from '@material-ui/core/styles/colorManipulator';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableFooter from '@material-ui/core/TableFooter';

import TableWidgetSelector, { getTableWidgetTags } from './widgetSelector';

const styles = theme => ({
  headerLayout: {
    tableLayout: 'fixed',
    width: 'calc(100% - 15px)',
  },
  headerLayoutNoScroll: {
    tableLayout: 'fixed',
    width: '100%',
  },
  tableLayout: {
    tableLayout: 'fixed',
  },
  tableBody: {
    overflowY: 'auto',
    height: 'calc(100% - 75px)',
    width: '100%',
  },
  rowFormat: {
    height: '30px',
  },
  textHeadings: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.3),
    textAlign: 'Center',
  },
  textBody: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.1),
    textAlign: 'Center',
    padding: '2px',
  },
});

const WidgetTable = props => {
  const values =
    props.localState === undefined
      ? props.attribute.value
      : props.localState.value;
  const columnLabels = Object.keys(props.attribute.meta.elements);
  const rowChangeHandler = (rowPath, newValue) => {
    const rowValue = {};
    columnLabels.forEach(label => {
      rowValue[label] = values[label][rowPath.row];
      return 0;
    });
    rowValue[rowPath.label] = newValue;
    props.eventHandler(props.attribute.path, rowValue, rowPath.row);
  };

  const columnWidgetTags = getTableWidgetTags(props.attribute);
  const rowNames = values[columnLabels[0]];
  const columnHeadings = columnLabels.map((label, column) => (
    <TableCell
      className={props.classes.textHeadings}
      padding="none"
      key={column}
    >
      {label}
    </TableCell>
  ));
  return (
    <div>
      <Table
        className={
          rowNames.length > 20
            ? props.classes.headerLayout
            : props.classes.headerLayoutNoScroll
        }
      >
        <TableHead>
          <TableRow className={props.classes.rowFormat}>
            {columnHeadings}
          </TableRow>
        </TableHead>
      </Table>
      <div className={props.classes.tableBody}>
        <Table className={props.classes.tableLayout}>
          <TableBody>
            {rowNames.map((name, row) => (
              <TableRow className={props.classes.rowFormat} key={row}>
                {columnLabels.map((label, column) => (
                  <TableCell
                    className={props.classes.textBody}
                    padding="none"
                    key={[row, column]}
                  >
                    <TableWidgetSelector
                      columnWidgetTag={columnWidgetTags[column]}
                      value={values[label][row]}
                      rowPath={{ label, row, column }}
                      rowChangeHandler={rowChangeHandler}
                      columnMeta={props.attribute.meta.elements[label]}
                      setFlag={props.setFlag}
                    />
                  </TableCell>
                ))}
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </div>
      <Table className={props.classes.headerLayout}>
        <TableFooter>
          <TableRow className={props.classes.rowFormat}>
            {props.footerItems.map((item, key) => (
              <TableCell key={key}>{item}</TableCell>
            ))}
          </TableRow>
        </TableFooter>
      </Table>
    </div>
  );
};

WidgetTable.propTypes = {
  attribute: PropTypes.shape({
    path: PropTypes.arrayOf(PropTypes.string),
    value: PropTypes.shape({}),
    name: PropTypes.string,
    typeid: PropTypes.string,
    labels: PropTypes.arrayOf(PropTypes.string),
    pending: PropTypes.bool,
    errorState: PropTypes.bool,
    alarm: PropTypes.shape({
      severity: PropTypes.number,
    }),
    meta: PropTypes.shape({
      label: PropTypes.string,
      tags: PropTypes.arrayOf(PropTypes.string),
      elements: PropTypes.shape({}),
    }),
    unableToProcess: PropTypes.bool,
  }).isRequired,
  localState: PropTypes.shape({
    value: PropTypes.shape({}),
  }),
  classes: PropTypes.shape({
    tableBody: PropTypes.string,
    headerLayoutNoScroll: PropTypes.string,
    headerLayout: PropTypes.string,
    tableLayout: PropTypes.string,
    textHeadings: PropTypes.string,
    textBody: PropTypes.string,
    rowFormat: PropTypes.string,
  }).isRequired,
  eventHandler: PropTypes.func.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  setFlag: PropTypes.func.isRequired,
  footerItems: PropTypes.arrayOf(PropTypes.node),
};

WidgetTable.defaultProps = {
  localState: undefined,
  footerItems: [],
};
export default withStyles(styles, { withTheme: true })(WidgetTable);
