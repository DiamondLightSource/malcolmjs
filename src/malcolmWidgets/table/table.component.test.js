import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import WidgetTable from './table.component';
import { harderAttribute } from './table.stories';

describe('WidgetTable', () => {
  let shallow;
  let mount;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    mount = createMount();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <WidgetTable
        attribute={harderAttribute}
        eventHandler={() => {}}
        setFlag={() => {}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls set flag on click', () => {
    const setFlag = jest.fn();
    const wrapper = mount(
      <WidgetTable
        attribute={harderAttribute}
        eventHandler={() => {}}
        setFlag={setFlag}
      />
    );
    wrapper
      .find('input')
      .first()
      .simulate('focus');
    expect(setFlag.mock.calls.length).toEqual(1);
    expect(setFlag.mock.calls[0]).toEqual([
      { column: 2, label: 'x', row: 0 },
      'dirty',
      true,
    ]);
  });

  it('calls set value on checkbox click', () => {
    const setValue = jest.fn();
    const wrapper = mount(
      <WidgetTable
        attribute={harderAttribute}
        eventHandler={setValue}
        setFlag={() => {}}
      />
    );
    wrapper
      .find('input')
      .last()
      .simulate('change');
    expect(setValue.mock.calls.length).toEqual(1);
    expect(setValue.mock.calls[0]).toEqual([
      ['PANDA', 'layout'],
      { mri: 'PANDA:INENC2', name: 'INENC2', visible: false, x: 0, y: 0 },
    ]);
  });
});
