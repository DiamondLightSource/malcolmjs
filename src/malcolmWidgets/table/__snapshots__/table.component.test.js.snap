// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetTable renders correctly 1`] = `
<div>
  <WithStyles(Table)
    className="WidgetTable-headerLayoutNoScroll-2"
  >
    <WithStyles(TableHead)>
      <WithStyles(TableRow)
        className="WidgetTable-rowFormat-5"
      >
        <WithStyles(TableCell)
          className="WidgetTable-textHeadings-6"
          key="0"
          padding="none"
        >
          name
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          className="WidgetTable-textHeadings-6"
          key="1"
          padding="none"
        >
          mri
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          className="WidgetTable-textHeadings-6"
          key="2"
          padding="none"
        >
          x
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          className="WidgetTable-textHeadings-6"
          key="3"
          padding="none"
        >
          y
        </WithStyles(TableCell)>
        <WithStyles(TableCell)
          className="WidgetTable-textHeadings-6"
          key="4"
          padding="none"
        >
          visible
        </WithStyles(TableCell)>
      </WithStyles(TableRow)>
    </WithStyles(TableHead)>
  </WithStyles(Table)>
  <div
    className="WidgetTable-tableBody-4"
  >
    <WithStyles(Table)
      className="WidgetTable-tableLayout-3"
    >
      <WithStyles(TableBody)>
        <WithStyles(TableRow)
          className="WidgetTable-rowFormat-5"
          key="0"
        >
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="0,0"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Names of the layout parts",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 0,
                  "label": "name",
                  "row": 0,
                }
              }
              setFlag={[Function]}
              value="TTLIN1"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="0,1"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Malcolm full names of child blocks",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 1,
                  "label": "mri",
                  "row": 0,
                }
              }
              setFlag={[Function]}
              value="PANDA:TTLIN1"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="0,2"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "X Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 2,
                  "label": "x",
                  "row": 0,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="0,3"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Y Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 3,
                  "label": "y",
                  "row": 0,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="0,4"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Whether child blocks are visible",
                  "label": "",
                  "tags": Array [
                    "widget:checkbox",
                  ],
                  "typeid": "malcolm:core/BooleanArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:checkbox"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 4,
                  "label": "visible",
                  "row": 0,
                }
              }
              setFlag={[Function]}
              value={false}
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          className="WidgetTable-rowFormat-5"
          key="1"
        >
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="1,0"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Names of the layout parts",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 0,
                  "label": "name",
                  "row": 1,
                }
              }
              setFlag={[Function]}
              value="TTLIN2"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="1,1"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Malcolm full names of child blocks",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 1,
                  "label": "mri",
                  "row": 1,
                }
              }
              setFlag={[Function]}
              value="PANDA:TTLIN2"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="1,2"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "X Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 2,
                  "label": "x",
                  "row": 1,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="1,3"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Y Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 3,
                  "label": "y",
                  "row": 1,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="1,4"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Whether child blocks are visible",
                  "label": "",
                  "tags": Array [
                    "widget:checkbox",
                  ],
                  "typeid": "malcolm:core/BooleanArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:checkbox"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 4,
                  "label": "visible",
                  "row": 1,
                }
              }
              setFlag={[Function]}
              value={false}
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          className="WidgetTable-rowFormat-5"
          key="2"
        >
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="2,0"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Names of the layout parts",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 0,
                  "label": "name",
                  "row": 2,
                }
              }
              setFlag={[Function]}
              value="INENC1"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="2,1"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Malcolm full names of child blocks",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 1,
                  "label": "mri",
                  "row": 2,
                }
              }
              setFlag={[Function]}
              value="PANDA:INENC1"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="2,2"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "X Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 2,
                  "label": "x",
                  "row": 2,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="2,3"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Y Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 3,
                  "label": "y",
                  "row": 2,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="2,4"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Whether child blocks are visible",
                  "label": "",
                  "tags": Array [
                    "widget:checkbox",
                  ],
                  "typeid": "malcolm:core/BooleanArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:checkbox"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 4,
                  "label": "visible",
                  "row": 2,
                }
              }
              setFlag={[Function]}
              value={false}
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          className="WidgetTable-rowFormat-5"
          key="3"
        >
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="3,0"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Names of the layout parts",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 0,
                  "label": "name",
                  "row": 3,
                }
              }
              setFlag={[Function]}
              value="INENC2"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="3,1"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Malcolm full names of child blocks",
                  "label": "",
                  "tags": Array [
                    "widget:textupdate",
                  ],
                  "typeid": "malcolm:core/StringArrayMeta:1.0",
                  "writeable": false,
                }
              }
              columnWidgetTag="widget:textupdate"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 1,
                  "label": "mri",
                  "row": 3,
                }
              }
              setFlag={[Function]}
              value="PANDA:INENC2"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="3,2"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "X Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 2,
                  "label": "x",
                  "row": 3,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="3,3"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Y Coordinates of child blocks",
                  "dtype": "float64",
                  "label": "",
                  "tags": Array [
                    "widget:textinput",
                  ],
                  "typeid": "malcolm:core/NumberArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:textinput"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 3,
                  "label": "y",
                  "row": 3,
                }
              }
              setFlag={[Function]}
              value={0}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            className="WidgetTable-textBody-7"
            key="3,4"
            padding="none"
          >
            <WithTheme(TableWidgetSelector)
              columnMeta={
                Object {
                  "description": "Whether child blocks are visible",
                  "label": "",
                  "tags": Array [
                    "widget:checkbox",
                  ],
                  "typeid": "malcolm:core/BooleanArrayMeta:1.0",
                  "writeable": true,
                }
              }
              columnWidgetTag="widget:checkbox"
              rowChangeHandler={[Function]}
              rowPath={
                Object {
                  "column": 4,
                  "label": "visible",
                  "row": 3,
                }
              }
              setFlag={[Function]}
              value={false}
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </div>
  <WithStyles(Table)
    className="WidgetTable-headerLayout-1"
  >
    <WithStyles(TableFooter)>
      <WithStyles(TableRow)
        className="WidgetTable-rowFormat-5"
      />
    </WithStyles(TableFooter)>
  </WithStyles(Table)>
</div>
`;
